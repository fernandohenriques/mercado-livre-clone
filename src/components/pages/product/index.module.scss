.container {
  height: auto;
  min-height: 100%;
  width: 80%;

  .product {
    padding: 0 1rem;
  }
}
