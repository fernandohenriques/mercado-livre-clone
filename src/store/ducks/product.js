import Api from 'services/api';

const initialState = {
  loading: false,
  error: false,
  products: [{
    id: 0,
    title: '',
    description: '',
    picture: '',
    price: {
      amount: 0,
    },
    sold_quantity: 0,
    free_shipping: false,
  }],
};

const actionTypes = {
  FETCH_PRODUCT_PENDING: 'app/product/FETCH_PRODUCT_PENDING',
  FETCH_PRODUCT_SUCCESS: 'app/product/FETCH_SEARCH_SUCCESS',
  FETCH_PRODUCT_ERROR: 'app/product/FETCH_SEARCH_ERROR',
};

const { FETCH_PRODUCT_PENDING, FETCH_PRODUCT_SUCCESS, FETCH_PRODUCT_ERROR } = actionTypes;

const fetchProductPending = () => ({
  type: FETCH_PRODUCT_PENDING,
});

const fetchProductSuccess = (payload) => ({
  type: FETCH_PRODUCT_SUCCESS,
  payload,
});

const fetchProductError = () => ({
  type: FETCH_PRODUCT_ERROR,
});

const reducer = (state = initialState, action) => {
  switch (action.type) {
  case FETCH_PRODUCT_PENDING: {
    return {
      ...state,
      loading: true,
    };
  }
  case FETCH_PRODUCT_SUCCESS: {
    const product = action.payload;
    return {
      error: false,
      loading: false,
      products: [...state.products, product],
    };
  }
  case FETCH_PRODUCT_ERROR: {
    return {
      ...state,
      error: true,
      loading: false,
    };
  }
  default:
    return state;
  }
};

const fetchProduct = (id) => {
  const { getProduct } = new Api();

  return async (dispatch, getState) => {
    const { products } = getState();
    const product = products.find((item) => item.id === id);

    if (product) return;

    dispatch(fetchProductPending());

    try {
      const result = await getProduct(id);
      dispatch(fetchProductSuccess(result));
    } catch (e) {
      dispatch(fetchProductError());
    }
  };
};

export { initialState, fetchProduct };
export default reducer;
