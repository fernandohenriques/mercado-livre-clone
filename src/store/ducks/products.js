const initialState = [
  {
    id: 0,
    title: '',
    picture: '',
    price: {
      amount: 0,
    },
    adress: {
      state_name: '',
    },
    free_shipping: false,
  },
];

const actionTypes = {
  SET_PRODUCTS: 'SET_PRODUCTS',
};

const setProducts = (products) => ({
  type: 'SET_PRODUCTS',
  products,
});

const reducer = (state = initialState, action) => {
  switch (action.type) {
  case actionTypes.SET_PRODUCTS:
    // console.log(state);
    // console.log(action);

    return state;
  default:
    return state;
  }
};

export { actionTypes, setProducts };
export default reducer;
