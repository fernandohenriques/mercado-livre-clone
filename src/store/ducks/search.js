import Api from 'services/api';

const initialState = {
  loading: false,
  error: false,
  result: {
    categories: [],
    products: [],
  },
};

const actionTypes = {
  FETCH_SEARCH_PENDING: 'app/search/FETCH_SEARCH_PENDING',
  FETCH_SEARCH_SUCCESS: 'app/search/FETCH_SEARCH_SUCCESS',
  FETCH_SEARCH_ERROR: 'app/search/FETCH_SEARCH_ERROR',
};

const { FETCH_SEARCH_PENDING, FETCH_SEARCH_SUCCESS, FETCH_SEARCH_ERROR } = actionTypes;

const fetchSearchPending = () => ({
  type: FETCH_SEARCH_PENDING,
});

const fetchSearchSuccess = (payload) => ({
  type: FETCH_SEARCH_SUCCESS,
  payload,
});

const fetchSearchError = () => ({
  type: FETCH_SEARCH_ERROR,
});

const reducer = (state = initialState, action) => {
  switch (action.type) {
  case FETCH_SEARCH_PENDING: {
    return {
      ...state,
      loading: true,
    };
  }
  case FETCH_SEARCH_SUCCESS: {
    const result = {
      categories: action.payload.categories || [],
      products: action.payload.items || [],
    };
    return {
      ...state,
      result,
      error: false,
      loading: false,
    };
  }
  case FETCH_SEARCH_ERROR: {
    return {
      ...state,
      error: true,
      loading: false,
    };
  }
  default:
    return state;
  }
};

const fetchSearch = (q) => {
  const { getProducts } = new Api();

  return async (dispatch) => {
    dispatch(fetchSearchPending());

    try {
      const result = await getProducts(q);
      dispatch(fetchSearchSuccess(result));
    } catch (e) {
      dispatch(fetchSearchError());
    }
  };
};

export { initialState, fetchSearch };
export default reducer;
