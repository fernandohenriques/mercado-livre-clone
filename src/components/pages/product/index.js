import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Button from 'components/atoms/button';
import Page from 'components/templates/page';
import Loading from 'components/atoms/loading';
import SuperCard from 'components/atoms/superCard';
import Breadcrumbs from 'components/atoms/breadcrumbs';
import { fetchProduct as fetchProductAction } from 'store/ducks/product';

import styles from './index.module.scss';

const Product = ({ loading, product, categories }) => {
  const conditions = new Map([
    ['new', 'Nuevo'],
    ['not_specified', 'No especificado'],
    ['used', 'Usado'],
  ]);

  const hasSoldQuantity = typeof product.sold_quantity !== 'undefined';

  return (
    <Page title={`Mercado Livre - ${product ? product.title : 'Produto não encontrado'}`}>
      <div className={styles.container}>
        {loading ? <Loading /> : (
          <>
            <Breadcrumbs categories={categories} />
            <SuperCard className={styles.product}>
              {product ? (
                <>
                  <section className={styles.coverAndData}>
                    <div className={styles.cover}>
                      <img src={product.picture} alt={product.title} title={product.title} />
                    </div>
                    <div className={styles.data}>
                      <div className={styles.condition}>
                        {`${conditions.get(product.condition)}${hasSoldQuantity ? ` - ${product.sold_quantity} vendidos` : ''}`}
                      </div>
                      <div className={styles.name}>
                        <strong>{product.title}</strong>
                      </div>
                      <div className={styles.price}>
                        <strong>{`$ ${product.price.amount.toFixed(2)}`}</strong>
                      </div>
                      <Button onClick={() => null}>
                        Comprar
                      </Button>
                    </div>
                  </section>
                  {product.description && (
                    <section className={styles.description}>
                      <span>Descripción del producto</span>
                      <p>{product.description}</p>
                    </section>
                  )}
                </>
              ) : <h3>Produto não encontrado.</h3>}
            </SuperCard>
          </>
        )}
      </div>
    </Page>
  );
};

Product.getInitialProps = ({ store, query: { id } }) => {
  if (id && id !== '') return store.dispatch(fetchProductAction(id));

  return {};
};

Product.propTypes = {
  loading: PropTypes.bool,
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    price: PropTypes.shape({
      currency: PropTypes.string,
      amount: PropTypes.number.isRequired,
      decimals: PropTypes.number,
    }).isRequired,
    picture: PropTypes.string.isRequired,
    condition: PropTypes.string.isRequired,
    free_shipping: PropTypes.bool.isRequired,
    sold_quantity: PropTypes.number,
  }),
  categories: PropTypes.arrayOf(PropTypes.string.isRequired),
};

Product.defaultProps = {
  loading: false,
  product: null,
  categories: [],
};

const mapStateToProps = ({ product: { loading, lastIdfetched, products }, search: { result } }) => ({
  loading,
  product: products.find((item) => item.id === lastIdfetched),
  categories: result.categories,
});

export default connect(mapStateToProps)(Product);
