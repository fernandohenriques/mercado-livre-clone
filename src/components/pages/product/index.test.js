import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { render } from '@testing-library/react';

import { initialState as searchInitialState } from 'store/ducks/search';
import { initialState as productInisialState } from 'store/ducks/product';
import Product from './index';

describe('pages <Product />', () => {
  const mockStore = configureMockStore();
  const store = mockStore({ search: searchInitialState, product: productInisialState });

  it('Snapshot testing', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Product />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
