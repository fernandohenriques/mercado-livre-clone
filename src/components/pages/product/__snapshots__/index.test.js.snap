// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages <Product /> Snapshot testing 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <div
        class="logo"
        role="button"
        tabindex="0"
      >
        <figure>
          <img
            alt="Mercado Livre - Onde comprar e vender de Tudo"
            src="logo.png"
            title="Mercado Livre - Onde comprar e vender de Tudo"
          />
        </figure>
      </div>
      <div
        class="container"
      >
        <input
          class="input input"
          placeholder="Buscar produtos..."
          type="text"
          value=""
        />
        <button
          type="button"
        >
          <img
            alt="Buscar"
            src="ic-search.png"
            title="Buscar"
          />
        </button>
      </div>
    </div>
  </header>
  <main>
    <div
      class="container"
    >
      <div
        class="container"
      />
      <div
        class="container product"
      >
        <section
          class="coverAndData"
        >
          <div
            class="cover"
          >
            <img
              alt=""
              src=""
              title=""
            />
          </div>
          <div
            class="data"
          >
            <div
              class="condition"
            >
              Nuevo - 0 vendidos
            </div>
            <div
              class="name"
            >
              <strong />
            </div>
            <div
              class="price"
            >
              <strong>
                <span
                  class="price large"
                >
                  <span
                    class="amount"
                  >
                    $ 0
                  </span>
                </span>
              </strong>
            </div>
            <button
              class="button"
              title="Comprar"
              type="button"
            >
              Comprar
            </button>
          </div>
        </section>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
